src/tools/diagnose/analyze_resource/analyze_resource.js

```diff
@@ -0,0 +1,140 @@
+/* Copyright (C) 2023 NooBaa */
+'use strict';
+
+const dbg = require('../../../util/debug_module')(__filename);
+dbg.set_process_name('analyze_resource');
+
+const get_cloud_vendor = require('./analyze_resource_cloud_vendor').get_cloud_vendor;
+
+const connection_basic_details = {
+    bucket: process.env.BUCKET,
+    endpoint: process.env.ENDPOINT,
+    signatureVersion: process.env.RESOURCE_TYPE === 'aws-s3' ? 'v4' : process.env.S3_SIGNATURE_VERSION,
+};
+
+const testing_status = {
+    pass: [],
+    fail: [],
+    skip: [],
+};
+
+async function main() {
+    try {
+        const resource_type = process.env.RESOURCE_TYPE;
+        check_supported_resource_type(resource_type);
+        const cloud_vendor = get_cloud_vendor(resource_type, connection_basic_details);
+        await test_resource(cloud_vendor, connection_basic_details.bucket);
+    } catch (err) {
+        console.error(`Test Diagnose Resource Failed: ${err}`);
+        process.exit(1);
+    }
+    process.exit(0);
+}
+
+function check_supported_resource_type(resource_type) {
+    const SUPPORTED_TYPES_OF_RESOURCES = ['aws-s3', 's3-compatible', 'ibm-cos', 'google-cloud-storage', 'azure-blob']; //'pv-pool' and 'nsfs' are not cloud vendor and we will not check them
+    if (SUPPORTED_TYPES_OF_RESOURCES.includes(resource_type)) {
+        console.info('Diagnose Resource Started');
+    } else {
+        throw new Error(`${resource_type} is not supported`);
+    }
+}
+
+async function test_resource(cloud_vendor, bucket) {
+    check_arguments(cloud_vendor, bucket);
+    console.info(`We will have several tests to check the status of the resource and permission of the bucket ${bucket}`);
+    const key_to_head = await list_objects_and_get_one_key(cloud_vendor, bucket);
+    if (key_to_head) {
+        console.info(`A returned key: ${key_to_head}`);
+        await head_object(cloud_vendor, bucket, key_to_head);
+    } else {
+        testing_status.skip.push('head object');
```

Comment:
in case the bucket is empty, then maybe head test should be done with the key that is uploaded later. only if there is also no write permissions, then we can't perform the head test.
also, it's worth printing to log why the test was skipped